### PR TITLE
Refactor slicer steps

### PR DESCRIPTION
**Pull request by letFunny**
- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

### CODE REVIEW

File: internal/slicer/slicer.go [hunk 1]
```diff
@@ -358,11 +271,108 @@ func Run(options *RunOptions) (*Report, error) {
...
+			err = fmt.Errorf("cannot read file which is not selected: %s", path)
 		}
 	}
-	return false
+	return err
+}
+
+// TODO comment.
+func addParents(knownPaths map[string]bool, path string) {
+	if path[0] != '/' {
```
Comment (rebornplusplus) [inline internal/slicer/slicer.go:298]: nit: `strings.HasPrefix`?

File: internal/slicer/slicer.go [hunk 2]
```diff
@@ -358,11 +271,108 @@ func Run(options *RunOptions) (*Report, error) {
...
+		panic("bug: tried to add relative path to known paths")
+	}
+	cleanPath := filepath.Clean(path)
+	slashPath := cleanPath
+	if path[len(path)-1] == '/' && cleanPath != "/" {
```
Comment (rebornplusplus) [inline internal/slicer/slicer.go:303]: nit: `strings.HasSuffix`?

File: internal/slicer/slicer.go (hunk 1 above)
Comment (rebornplusplus) [reply internal/slicer/slicer.go:298]: Also do you wanna handle `path == ""`?

File: internal/slicer/slicer.go [hunk 3]
```diff
@@ -358,11 +271,108 @@ func Run(options *RunOptions) (*Report, error) {
...
+		reader, err := archive.Fetch(slice.Package)
+		if err != nil {
+			return nil, err
+		}
+		packages[slice.Package] = reader
+	}
+	return packages, nil
+}
+
+func createFileFromPathInfo(targetPath string, pathInfo setup.PathInfo) (*fsutil.Entry, error) {
```
Comment (rebornplusplus) [inline internal/slicer/slicer.go:344]: nit: `createFile` or `createFromPathInfo`?

File: internal/slicer/slicer.go [hunk 4]
```diff
@@ -284,27 +214,10 @@ func Run(options *RunOptions) (*Report, error) {
...
-					err = fmt.Errorf("cannot list directory which is not selected: %s", path)
-				}
-			} else {
-				if knownPaths[path+"/"] {
-					err = fmt.Errorf("content is not a file: %s", path)
-				} else {
-					err = fmt.Errorf("cannot read file which is not selected: %s", path)
-				}
-			}
+			return checkReadError(knownPaths, path)
```
Comment (rebornplusplus) [inline internal/slicer/slicer.go:218]: Hmmm.... I see how `checkReadError` is used now. The function alone (down) seemed weird to me at first, not knowing where this will be used.
Maybe it could use more comments, or instead creating a new function, keeping the content inlined as before. The issue with the latter approach is that the `slicer.Run` function will only grow in size.

Maybe running the mutation scripts should be in it's own function?

File: internal/slicer/slicer.go
Comment (rebornplusplus) [inline internal/slicer/slicer.go:1]: What you are doing here is quite nice! Thanks.

If I were doing this, I would probably try to keep as less stuff in `slicer.Run` as possible, divide it among the following functions/blocks and only call them from `slicer.Run`:
* Fetch packages.
* Build info to process selection.
* Extract.
* Create the contents not coming from package. (maybe this one can be inside extract too)
* Run mutation scripts.
* Remove the `until` paths.

Basically, I would take each block of code and put it in it's own function. `slicer.Run` would only call some function, take some info, pass it onto something else, do XYZ... and so on..

**Review by rebornplusplus — APPROVED** [review]
Looks good to me, thank you. Nice changes, looks better in most cases!

File: internal/slicer/slicer.go
Comment (letFunny) [reply internal/slicer/slicer.go:1]: That was my first instinct as well but the code gets messy because there is a lot of shared state so it needs to call a function with 5 arguments, 2 of which are mutated and then call the next one with another set of arguments and on and on.

My first instinct was that the style we follow in Chisel does not disallow long functions when state is interleaved (see extract.go) so instead what I tried to do was create functions to move the parts that are not very readable (e.g. create the file based on the tar permissions) and are not needed for the overall function.

File: internal/slicer/slicer.go (hunk 4 above)
Comment (letFunny) [reply internal/slicer/slicer.go:218]: Good suggestion, I will try to make it its own function and see how it looks. My first impressions was that the code was not too lengthy and that we would need to pass a lot of variables (`knownPaths`, `pathInfos`, `slice`, etc.)

**Comment by letFunny** [comment]
https://github.com/canonical/chisel/pull/136
